### PR TITLE
Feat: add support for files

### PR DESCRIPTION
**Pull request by eoaksnes**
## What does this pull request change?

* Added `EmployeeApp` that demonstrates files (of `File` blueprint) that are uploaded using DM CLI
* Added a prototype of a `File` plugin that can be used to upload and download files
* Added a prototype `BinaryField` for the form plugin that can be used to download binary data

> **_NOTE:_**   Did not spend too much time on the `File` plugin or the `BinaryField` field of the form, since the functionality is not agreed on, so this is to just be able to upload and download files. Should be done better when gone is used in real applications.

### CODE REVIEW

File: packages/dm-core-plugins/src/form/components/OpenObjectButton.tsx [hunk 1]
```diff
@@ -2,17 +2,29 @@ import { Button } from '@equinor/eds-core-react'
...
   return (
     <Button
       variant="outlined"
       onClick={() =>
-        onOpen?.(namePath, {
-          type: 'ViewConfig',
-          scope: namePath,
-        })
+        onOpen?.(
+          viewId || namePath,
```
Comment (KristianKjerstad) [inline packages/dm-core-plugins/src/form/components/OpenObjectButton.tsx:21]: are both viewId and namePath required props at the same time? Or should only one be passed in at a time?

If the second case is true, the correct type would be:

```TypeScript
type Props =
      ({
          viewId: string
          namePath?: never
        }
      | {
          viewId?: never;
          namePath: string;
        })
      & { idReference?: string };
```
This way, either viewId or namePath, but not both.

File: packages/dm-core-plugins/src/file/FilePlugin.tsx [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+export const ErrorGroup = styled.div`
+  display: flex;
+  flex-direction: column;
+  border: 1px solid rgba(213, 18, 18, 0.71);
+  border-radius: 5px;
+  padding: 20px 20px;
+  background-color: #f6dfdf;
+`
+
+const UploadFileButton = (props: any) => {
```
Comment (KristianKjerstad) [inline packages/dm-core-plugins/src/file/FilePlugin.tsx:26]:
```suggestion
const UploadFileButton = (props: {onUpload: (file ?, reference: TLinkReference) => void}) => {
```

not sure what type of file is. 


Without adding type to props this plugin is very difficult to use

File: packages/dm-core-plugins/src/file/FilePlugin.tsx [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+        >
+          Upload
+        </Button>
+      )}
+      {error && <p style={{ color: 'red' }}>{error}</p>}
+    </>
+  )
+}
+
+const DownloadButton = (props: any) => {
```
Comment (KristianKjerstad) [inline packages/dm-core-plugins/src/file/FilePlugin.tsx:106]: DownloadButton should be moved to a new file DownloadButton.tsx

File: packages/dm-core-plugins/src/file/FilePlugin.tsx (hunk 2 above)
Comment (KristianKjerstad) [inline packages/dm-core-plugins/src/file/FilePlugin.tsx:26]: UploadFileButton should be moved to a new file UploadFileButton.tsx

File: packages/dm-core-plugins/src/file/FilePlugin.tsx (hunk 3 above)
Comment (KristianKjerstad) [inline packages/dm-core-plugins/src/file/FilePlugin.tsx:106]:
```suggestion
const DownloadButton = (props:{entity: TGenericObject, dataSourceId: string}) => {
```

File: packages/dm-core-plugins/src/file/FilePlugin.tsx [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+  const { entity, dataSourceId } = props
+  const [fileUrl, setFileUrl] = useState('')
```
Comment (KristianKjerstad) [inline packages/dm-core-plugins/src/file/FilePlugin.tsx:108]:
```suggestion
  const [fileUrl, setFileUrl] = useState<string>('')
```

File: packages/dm-core-plugins/src/file/FilePlugin.tsx [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+  const [loading, setLoading] = useState(true)
```
Comment (KristianKjerstad) [inline packages/dm-core-plugins/src/file/FilePlugin.tsx:109]:
```suggestion
  const [loading, setLoading] = useState<boolean>(true)
```

File: packages/dm-core-plugins/src/file/FilePlugin.tsx [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+  const [error, setError] = useState<string | null>(null)
+
+  const dmssAPI = useDMSS()
+
+  useEffect(() => {
+    setError(null)
```
Comment (KristianKjerstad) [inline packages/dm-core-plugins/src/file/FilePlugin.tsx:115]:
```suggestion
    setError(null)
    setLoading(true)
```

File: packages/dm-core-plugins/src/file/FilePlugin.tsx [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+    const options: AxiosRequestConfig = {
+      responseType: 'blob',
+    }
+    dmssAPI
+      .blobGetById(
+        {
+          dataSourceId: dataSourceId,
+          blobId: entity.content['address'],
```
Comment (KristianKjerstad) [inline packages/dm-core-plugins/src/file/FilePlugin.tsx:123]: `entity.content['address']` assumes that entity sent in as props is of a special type, but i'm not sure which. Is it a link reference?

Can rename entity prop to something like "LinkReferenceEntity" or whateter type its supposed to be

File: packages/dm-core-plugins/src/file/FilePlugin.tsx [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+        },
+        options
+      )
+      .then((response: any) => {
```
Comment (KristianKjerstad) [inline packages/dm-core-plugins/src/file/FilePlugin.tsx:127]:
```suggestion
      .then((response: AxiosResponse) => {
```

File: packages/dm-core-plugins/src/file/FilePlugin.tsx [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+      >
+        Download
+      </Button>
+    </div>
+  )
+}
+
+export const FilePlugin = (props: IUIPlugin) => {
+  const { idReference } = props
+  const [entity, loading, updateDocument, error] = useDocument<TGenericObject>(
```
Comment (KristianKjerstad) [inline packages/dm-core-plugins/src/file/FilePlugin.tsx:169]: Looks like entity must be of type CORE:File.

can add a check in this plugin and throw error if `entity["type"] !== "dmss://system/SIMOS/Blueprint"`

Comment (KristianKjerstad) [inline packages/dm-core-plugins/src/file/FilePlugin.tsx:169]:
```suggestion
  const [entity: fileEntity, loading, updateDocument, error] = useDocument<TGenericObject>(
```

File: packages/dm-core-plugins/src/form/fields/BinaryField.tsx [hunk 10]
```diff
@@ -0,0 +1,79 @@
+import React from 'react'
+import { useFormContext } from 'react-hook-form'
+import { TStringFieldProps } from '../types'
+import { Button, Label } from '@equinor/eds-core-react'
+import { useRegistryContext } from '../RegistryContext'
+import { ErrorResponse, useDMSS } from '@development-framework/dm-core'
+import { AxiosError } from 'axios'
+
+const getTarget = (initialValue: any) => {
```
Comment (KristianKjerstad) [inline packages/dm-core-plugins/src/form/fields/BinaryField.tsx:9]:
```suggestion
const getTarget = (initialValue: TGenericObject & { address: string }): [string, string] => {
```

File: packages/dm-core-plugins/src/form/fields/BinaryField.tsx [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+  const { idReference } = useRegistryContext()
+  if (initialValue['address'].includes('blob')) {
+    const address = initialValue['address'].replace('blob://', '')
+    return address.split('/', 2)
+  } else {
+    return [idReference?.split('/', 2)[0], initialValue['address']]
+  }
+}
+
+const DownloadBinary = (props: any) => {
```
Comment (KristianKjerstad) [inline packages/dm-core-plugins/src/form/fields/BinaryField.tsx:19]:
```suggestion
const DownloadBinary = (props: { fileId: string, namePath: string, displayLabel: string, initialValue:  TGenericObject & { address: string }}) => {
```

File: packages/dm-core-plugins/src/form/fields/BinaryField.tsx [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+  const dmssAPI = useDMSS()
+  const [data_source_id, blob_id] = getTarget(initialValue)
+
+  const handleDownload = () => {
+    dmssAPI
+      .blobGetById({
+        dataSourceId: data_source_id,
+        blobId: blob_id,
+      })
+      .then((response: any) => {
```
Comment (KristianKjerstad) [inline packages/dm-core-plugins/src/form/fields/BinaryField.tsx:30]:
```suggestion
      .then((response: AxiosResponse) => {
```

File: packages/dm-core-plugins/src/form/fields/BinaryField.tsx [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+  return (
+    <div>
+      <Label label={displayLabel} />
+      <Button
+        namePath={namePath}
+        label={displayLabel}
+        // eslint-disable-next-line @typescript-eslint/no-empty-function
+        onChange={() => {}}
+        onClick={() => handleDownload()}
+      >
```
Comment (KristianKjerstad) [inline packages/dm-core-plugins/src/form/fields/BinaryField.tsx:54]:
```suggestion
      <Button
        namePath={namePath}
        label={displayLabel}
        onClick={() => handleDownload()}
      >
```

cannot see why onChange is needed

File: packages/dm-core/src/components/TreeView.tsx [hunk 14]
```diff
@@ -53,6 +56,14 @@ const GetIcon = (props: { node: TreeNode; expanded: boolean }) => {
     return <label>{'{}'}</label>
   }
   switch (node.type) {
+    case EBlueprint.FILE: {
+      const fileColor = '#2965FF'
```
Comment (KristianKjerstad) [inline packages/dm-core/src/components/TreeView.tsx:60]: is `#2965FF`a color from EDS?

File: packages/dm-core-plugins/src/form/components/OpenObjectButton.tsx (hunk 1 above)
Comment (eoaksnes) [reply packages/dm-core-plugins/src/form/components/OpenObjectButton.tsx:21]: Both are required

File: packages/dm-core-plugins/src/file/FilePlugin.tsx (hunk 6 above)
Comment (eoaksnes) [reply packages/dm-core-plugins/src/file/FilePlugin.tsx:115]: loading is set to default true

File: example/app/data/DemoDataSource/apps/EmployeeApp/employee2.json [hunk 15]
```diff
@@ -0,0 +1,4 @@
+{
+  "type": "./blueprints/Employee",
+  "name": "employee2"
```
[the remark this answers is not in the document]
Comment (eoaksnes) [reply example/app/data/DemoDataSource/apps/EmployeeApp/employee2.json]: fxed

File: packages/dm-core-plugins/src/file/FilePlugin.tsx (hunk 2 above)
Comment (eoaksnes) [reply packages/dm-core-plugins/src/file/FilePlugin.tsx]: fixed

Comment (eoaksnes) [reply packages/dm-core-plugins/src/file/FilePlugin.tsx]: fixed

File: packages/dm-core-plugins/src/form/fields/BinaryField.tsx (hunk 10 above)
Comment (eoaksnes) [reply packages/dm-core-plugins/src/form/fields/BinaryField.tsx:13]: fixed

File: packages/dm-core-plugins/src/form/fields/BinaryField.tsx (hunk 11 above)
Comment (eoaksnes) [reply packages/dm-core-plugins/src/form/fields/BinaryField.tsx]: fixed

File: packages/dm-core-plugins/src/form/fields/BinaryField.tsx (hunk 12 above)
Comment (eoaksnes) [reply packages/dm-core-plugins/src/form/fields/BinaryField.tsx]: fixed

File: packages/dm-core-plugins/src/form/fields/BinaryField.tsx (hunk 13 above)
Comment (eoaksnes) [reply packages/dm-core-plugins/src/form/fields/BinaryField.tsx]: fixed

File: packages/dm-core/src/components/TreeView.tsx (hunk 14 above)
Comment (eoaksnes) [reply packages/dm-core/src/components/TreeView.tsx:60]: nope, but need to cleanup the tree later on, think this is okey for now

File: packages/dm-core-plugins/src/file/FilePlugin.tsx (hunk 7 above)
Comment (eoaksnes) [reply packages/dm-core-plugins/src/file/FilePlugin.tsx]: good idea, renamed to fileEntity

File: packages/dm-core-plugins/src/form/fields/BinaryField.tsx (hunk 10 above)
Comment (KristianKjerstad) [reply packages/dm-core-plugins/src/form/fields/BinaryField.tsx:13]: looks like type is still any

**Review by KristianKjerstad — APPROVED** [review]
good!

**Pull request merged**